tpm2/seal.go

```diff
@@ -0,0 +1,450 @@
+	}
+
+	return kd, nil
+}
+
+type createdPcrPolicyCounter struct {
+	tpm     *tpm2.TPMContext
+	session tpm2.SessionContext
+	pub     *tpm2.NVPublic
+}
+
+func (c *createdPcrPolicyCounter) Pub() *tpm2.NVPublic {
+	if c == nil {
+		return nil
+	}
+	return c.pub
+}
+
+func (c *createdPcrPolicyCounter) undefineOnError(err error) {
+	if c == nil {
+		return
+	}
+
+	if err == nil {
+		return
+	}
+
+	index, err := tpm2.CreateNVIndexResourceContextFromPublic(c.pub)
+	if err != nil {
+		return
+	}
+	c.tpm.NVUndefineSpace(c.tpm.OwnerHandleContext(), index, c.session)
+}
+
+// keyDataPolicyParams corresponds to the parameters of a key's computed authorization
+// policy, consisting of the digest algorithm, the policy digest and the associated policy
+// data.
+type keyDataPolicyParams struct {
+	Alg        tpm2.HashAlgorithmId
+	PolicyData keyDataPolicy
+	AuthPolicy tpm2.Digest
+}
+
+// makeKeyDataPolicy creates the policy data required to seal a key with makeKeyDataWithPolicy
+// and creates a PCR policy counter if required.
+func makeKeyDataPolicy(tpm *tpm2.TPMContext, pcrPolicyCounterHandle tpm2.Handle, authKey secboot.AuxiliaryKey,
+	session tpm2.SessionContext) (data *keyDataPolicyParams, pcrPolicyCounterOut *createdPcrPolicyCounter,
+	authKeyOut secboot.AuxiliaryKey, err error) {
+	// Create an auth key.
+	if authKey == nil {
+		authKey = make(secboot.AuxiliaryKey, 32)
+		if _, err := rand.Read(authKey); err != nil {
+			return nil, nil, nil, xerrors.Errorf("cannot create key for signing dynamic authorization policies: %w", err)
+		}
+	}
+	authPublicKey, err := newPolicyAuthPublicKey(authKey)
+	if err != nil {
+		return nil, nil, nil, xerrors.Errorf("cannot derive public area of key for signing dynamic authorization policies: %w", err)
+	}
+
+	// Create PCR policy counter, if requested.
+	var pcrPolicyCount uint64
+	var pcrPolicyCounter *createdPcrPolicyCounter
+	if pcrPolicyCounterHandle != tpm2.HandleNull {
+		if tpm == nil {
+			return nil, nil, nil, errors.New("cannot create a PCR policy counter without a TPM connection")
+		}
+
+		var pub *tpm2.NVPublic
+		pub, pcrPolicyCount, err = createPcrPolicyCounter(tpm, pcrPolicyCounterHandle, authPublicKey, session)
+		switch {
+		case tpm2.IsTPMError(err, tpm2.ErrorNVDefined, tpm2.CommandNVDefineSpace):
+			return nil, nil, nil, TPMResourceExistsError{pcrPolicyCounterHandle}
+		case isAuthFailError(err, tpm2.CommandNVDefineSpace, 1):
+			return nil, nil, nil, AuthFailError{tpm2.HandleOwner}
+		case err != nil:
+			return nil, nil, nil, xerrors.Errorf("cannot create new PCR policy counter: %w", err)
+		}
+
+		pcrPolicyCounter = &createdPcrPolicyCounter{
+			tpm:     tpm,
+			session: session,
+			pub:     pub}
+
+		defer func() { pcrPolicyCounter.undefineOnError(err) }()
+	}
+
+	alg := tpm2.HashAlgorithmSHA256
+
+	// Create the initial policy data
+	policyData, authPolicy, err := newKeyDataPolicy(alg, authPublicKey, pcrPolicyCounter.Pub(), pcrPolicyCount)
```

Comment:
I notice that the  pre-existing doc comment for newKeyDataPolicy doesn't say much about authPolicy ?